src/user.service.ts

```diff
@@ -14,76 +14,80 @@ import { Shop } from './shop.service';
 
 export class UserAddress {
 
-    constructor() {
-    }
-    name: string;
-    note: string;
-    floor: string;
-    streetAdress: string;
-    region: string;
-    postalCode: string;
-    primary: boolean;
-    geo: {
-      lat: number;
-      lng: number;
-    }
+  constructor() {
+  }
+  name: string;
+  note: string;
+  floor: string;
+  streetAdress: string;
+  region: string;
+  postalCode: string;
+  primary: boolean;
+  geo: {
+    lat: number;
+    lng: number;
+  }
 }
 
 export class UserCard {
 
-    constructor() {
-    }
-     type:string;
-     name:string;
-     number:string;
-     expiry:string;
-     provider:string;
-     alias:string;
-   }
-
-export class User {
-
   constructor() {
-    this.id = '';
-    this.displayName = '';
-    this.name = {
-      givenName: '',
-      familyName: '',
-    };
-    this.birthday = new Date();
-    this.gender = '';
-    this.tags = [];
-    this.url = '';
-
-    this.email = {
-      address: '',
-      cc: '',
-      status: ''
-    };
-
-    this.reminder = {
-      active: false,
-      weekdays: [],
-      time: null
-    };
-
-    this.roles = [];
-    this.shops = [];
-    this.provider = '';
-    this.url = '';
-
-    this.phoneNumbers = [{
-      number: '',
-      what: 'mobile'
-    }];
+  }
+  type: string;
+  name: string;
+  number: string;
+  expiry: string;
+  provider: string;
+  alias: string;
+}
 
-    this.photo = '';
+export class User {
 
-    this.logistic = {
-      postalCode: ''
-    };
+  constructor(json: any) {
+    if (json !== null) {
+      Object.assign(this, json);
+    } else {
+      this.id = '';
+      this.displayName = '';
+      this.name = {
+        givenName: '',
+        familyName: '',
+      };
+      this.birthday = new Date();
+      this.gender = '';
+      this.tags = [];
+      this.url = '';
+
+      this.email = {
+        address: '',
+        cc: '',
+        status: ''
+      };
+
+      this.reminder = {
+        active: false,
+        weekdays: [],
+        time: null
+      };
+
+      this.roles = [];
+      this.shops = [];
+      this.provider = '';
+      this.url = '';
+
+      this.phoneNumbers = [{
+        number: '',
+        what: 'mobile'
+      }];
+
+      this.photo = '';
+
+      this.logistic = {
+        postalCode: ''
+      };
```

Comment:
* on peut enlever le `if`
* ici aussi on doit avoir un Object.assign(this,defaultUser,json||{});
* initialiser les nested Object/Array, par exemple `geo={}, email={}, addresses=[], roles=[], shops=[], reminder={}, ...` enfin tout pour éviter les nested `?` dans le template
* je crois que seul phoneNumbers doit rester pré-initialiser? tu peux laisser en comment